src/libstd/io/buffered.rs

```diff
@@ -989,6 +989,68 @@ impl<W: Write> Write for LineWriter<W> {
         }
     }
 
+    // Vectored writes are very similar to the writes above, but adjusted for
+    // the list of buffers that we have to write.
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
+        if self.need_flush {
+            self.flush()?;
+        }
+
+        // Find the last newline, and failing that write the whole buffer
+        let last_newline = bufs
+            .iter()
+            .enumerate()
+            .rev()
+            .filter_map(|(i, buf)| {
+                let pos = memchr::memrchr(b'\n', buf)?;
+                Some((i, pos))
+            })
+            .next();
+        let (i, j) = match last_newline {
+            Some(pair) => pair,
+            None => return self.inner.write_vectored(bufs),
+        };
+        let (prefix, suffix) = bufs.split_at(i);
+        let (buf, suffix) = suffix.split_at(1);
+        let buf = &buf[0];
+
+        // Write everything up to the last newline, flushing afterwards. Note
+        // that only if we finished our entire `write_vectored` do we try the
+        // subsequent
+        // `write`
+        let mut n = 0;
+        let prefix_amt = prefix.iter().map(|i| i.len()).sum();
+        if prefix_amt > 0 {
+            n += self.inner.write_vectored(prefix)?;
+            self.need_flush = true;
+        }
+        if n == prefix_amt {
+            match self.inner.write(&buf[..=j]) {
```

Comment:
I'm not sure how I feel about one write_vectored call turning into up to `write_vectored, write, flush, write, write_vectored`. Maybe it makes sense for LineWriter, but IIRC in general we tend to want to keep write/write_vectored relative 1:1 with downstream calls.

Reply:
FWIW `LineWriter::write` is [already a call to `write`, `flush`, then `write`](https://github.com/rust-lang/rust/blob/2fee28e7138d8753487ed8895ce0f5f2e643ffad/src/libstd/io/buffered.rs#L957-L990). We could skip a few calls here by creating a local `Vec<IoSlice>`, but I figured that since it's all buffered here anyway a few more calls wouldn't matter too much.

The main thing is that we want to slice a `&[IoSlice]` as if it were `&[u8]` but we can't do that since we can't mutate `IoSlice` in place.

Reply:
Fair enough.